src/main/java/com/fasterxml/jackson/databind/node/NonRecursiveSerializer.java

```diff
@@ -0,0 +1,286 @@
+            serializeList(node, events, gen, provider, typeSer);
+        } else if (node.isObject()) {
+            final List<Object> events = new ArrayList<>();
+            events.add(START_OBJECT);
+            Map<String, JsonNode> nodeMap = ((ObjectNode) node)._contentsToSerialize(provider);
+            events.addAll(getDescendentNodes(nodeMap, provider, trimEmptyArray, skipNulls));
+            events.add(END_OBJECT);
+            serializeList(node, events, gen, provider, typeSer);
+        } else {
+            node.serialize(gen, provider);
+        }
+    }
+
+    private static void serializeList(final JsonNode rootNode, final List<Object> events,
+                                      final JsonGenerator gen, final SerializerProvider provider,
+                                      final TypeSerializer typeSer)
+            throws IOException {
+        final int size = events.size();
+        WritableTypeId typeIdDef = null;
+        Object ev;
+        for (int i = 0; i < size; i++) {
+            ev = events.get(i);
+            if (ev == START_ARRAY) {
+                if (typeSer != null && i == 0) {
+                    typeIdDef = typeSer.writeTypePrefix(gen, typeSer.typeId(rootNode, START_ARRAY));
+                } else {
+                    gen.writeStartArray();
+                }
+            } else if (ev == END_ARRAY) {
+                if (typeIdDef != null && i == (size - 1)) {
+                    typeSer.writeTypeSuffix(gen, typeIdDef);
+                } else {
+                    gen.writeEndArray();
+                }
+            } else if (ev == START_OBJECT) {
+                if (typeSer != null && i == 0) {
+                    typeIdDef = typeSer.writeTypePrefix(gen, typeSer.typeId(rootNode, START_OBJECT));
+                } else {
+                    gen.writeStartObject();
+                }
+            } else if (ev == END_OBJECT) {
+                if (typeIdDef != null && i == (size - 1)) {
+                    typeSer.writeTypeSuffix(gen, typeIdDef);
+                } else {
+                    gen.writeEndObject();
+                }
+            } else if (ev instanceof String) {
+                gen.writeFieldName((String) ev);
+            } else if (ev instanceof JsonNode) {
+                ((JsonNode) ev).serialize(gen, provider);
+            } else if (ev instanceof ContainerNodeWrapper) {
+                serializeList(rootNode, ((ContainerNodeWrapper) ev).getEvents(), gen, provider, null);
+            }
+        }
+    }
+
+    private static List<Object> getDescendentNodes(final JsonNode node, final SerializerProvider provider,
+                                                   final boolean trimEmptyArray, final boolean skipNulls) {
+        final LinkedList<ContainerNodeWrapper> containerNodeWrappers = new LinkedList<>();
+        final List<Object> events = new ArrayList<>();
+        final Stack<ContainerNodeWrapper> stack = new Stack<>();
```

Comment:
use Deque instead of Stack, it's faster

Reply:
Stack has synchronization blocks so a Deque like ArrayDeque makes sense. I'm not sure if ArrayList is any slower than any other List impl. It is the simplest and has good general purpose performance. Some List implementations are optimised for some operations but weak at others (LInkedLists are poor for iteration, for instance - but good for appends/removals at the head and tail).

Reply:
I was thinking ArrayDeque since it's almost a drop-in replacement for Stack, but ArrayList might be slightly faster for this purpose. Not sure if it matters though